fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -2664,31 +2717,35 @@ private boolean areAllRecordTypesSince(@Nullable Collection<RecordType> recordTy
                                                                                      @Nonnull Map<Index, List<RecordType>> indexes) {
         // Do this with the new indexes in write-only mode to avoid using one of them
         // when evaluating the snapshot record count.
-        MutableRecordStoreState writeOnlyState = recordStoreState.withWriteOnlyIndexes(indexes.keySet().stream().map(Index::getName).collect(Collectors.toList()));
+        MutableRecordStoreState writeOnlyState = recordStoreState.get().withWriteOnlyIndexes(indexes.keySet().stream().map(Index::getName).collect(Collectors.toList()));
         // If all the new indexes are only for a record type whose primary key has a type prefix, then we can scan less.
         RecordType singleRecordTypeWithPrefixKey = singleRecordTypeWithPrefixKey(indexes);
         if (singleRecordTypeWithPrefixKey != null) {
             // Get a count for just those records, either from a COUNT index on just that type or from a universal COUNT index grouped by record type.
-            MutableRecordStoreState saveState = recordStoreState;
+            MutableRecordStoreState saveState = recordStoreState.get();
             try {
-                recordStoreState = writeOnlyState;
+                while (!recordStoreState.compareAndSet(saveState, writeOnlyState)) {
+                    saveState = recordStoreState.get();
+                }
```

Comment:
I think this should probably just throw an error if there are concurrent modifications to the `recordStoreStateRef` at this time rather than retrying as the code doesn't properly handle merging the two states. Probably the same for the other instances setting the `recordStoreState` in this method.

Reply:
Yeah, probably this is not good enough. Throwing an error is not ok either. I could imagine a case where a parallel thread updates the state too. Not sure what the best action would be? Synchronizing this block completely may be too much of an overhead.

Reply:
Hm, and I think synchronizing this block also doesn’t quite get us what we want, either.

The idea behind throwing an error is that we are declaring that multiple threads updating the state at once is illegal behavior, which at the moment, I think we should consider it to be. (I mean, my original proposal was that we would only have update the state once, which I at least still think is probably for the best.)

It could be that the right approach is to remember the old states for each index rather than remembering the full state. Then, at the end of the index being updated, it should update just those states.

I think there’s still some weirdness related to parallel operations happening on the same index. At the moment, we can kind of run and hide and say we “don’t support that”, but at some point, we should consider making that better, too.

Reply:
Thinking thru it, I feel, the the act of faking the state here is really the problematic part. I think I agree with you that throwing an error is currently the best course of action.

Reply:
Aha, reading your comment again, I think you are proposing that we only do a set similar to what we used to do?

Reply:
I think my actual proposal was that instead of saving and setting the record store state, one could remember the old state. I think this is probably fine, though. It's a bit unfortunate actually that the record store state has to be mocked up like that, and I was going through in my head how one could rearchitect things to not need to. I think there might be a way where one could mark all new indexes as not to be built (at first), and then maybe build them if needed?

This probably suffices for this PR. I'm not super happy with the way `checkVersion` works in general in that it's kind of weird that one has to call it exactly once to get the right semantics. I'm going to keep thinking on how we could make this better.

Reply:
Sounds good. Thanks.